### PR TITLE
[Bug]: Pimcore 11 errors with missing class directories

### LINKED ISSUE #15413: [Bug]: Pimcore 11 errors with missing class directories
### Pimcore version

11.x (branch)

### Steps to reproduce

Install the latest pimcore version (11.x-dev).
Be sure that the folder `PIMCORE_CLASS_DIRECTORY . '/DataObject/Objectbrick/Data'` (`PIMCORE_OBJECTBRICK_CLASS_DIRECTORY`) is not there.
Execute `bin/console` for example, and you will get an error from `scandir`.
The problem also exists for field collection directory.

### Expected / actual behavior

The problem is caused by commit [ba76382](https://github.com/pimcore/pimcore/commit/ba76382a54bb46a14d42619f700b3a3f1aaf75b1).

### PR DESCRIPTION
**Pull request by robertSt7**
resolves #15413

### CODE REVIEW

File: lib/Maintenance/Tasks/TaskHelper.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
+<?php
```
Comment (blankse) [inline lib/Maintenance/Tasks/TaskHelper.php]:
```suggestion
<?php
declare(strict_types=1);
```

File: lib/Maintenance/Tasks/CleanupBrickTablesTask.php [hunk 2]
```diff
@@ -31,21 +31,23 @@ class CleanupBrickTablesTask implements TaskInterface
...
     public function __construct(LoggerInterface $logger)
     {
         $this->logger = $logger;
-
-        $files = array_diff(scandir(self::PIMCORE_OBJECTBRICK_CLASS_DIRECTORY), ['..', '.']);
-        foreach ($files as $file) {
-            $classname = str_replace('.php', '', $file);
-            $this->mapLowerToCamelCase[strtolower($classname)] = $classname ;
-        }
+        $this->helper = TaskHelper::create($logger);
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/CleanupBrickTablesTask.php]: why not implement it as a service and inject it here?

File: lib/Maintenance/Tasks/CleanupFieldcollectionTablesTask.php [hunk 3]
```diff
@@ -19,32 +19,36 @@
...
     public function __construct(LoggerInterface $logger)
     {
         $this->logger = $logger;
-
-        $files = array_diff(scandir(self::PIMCORE_FIELDCOLLECTION_CLASS_DIRECTORY), ['..', '.']);
-        foreach ($files as $file) {
-            $classname = str_replace('.php', '', $file);
-            $this->mapLowerToCamelCase[strtolower($classname)] = $classname;
-        }
+        $this->helper = TaskHelper::create($logger);
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/CleanupFieldcollectionTablesTask.php]: inject it here?

File: lib/Maintenance/Tasks/TaskHelper.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        if (!self::$instance) {
+            self::$instance = new self();
+        }
+        self::$instance->logger = $logger;
+
+        return self::$instance;
+    }
+    
+    public function getDataStructureNamesMapLowerToCamelCase(string $dir): array
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/TaskHelper.php]:
```suggestion
    public function getCollectionNames(string $dir): array
```
I don't see any camel case conversion.

File: lib/Maintenance/Tasks/TaskHelper.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        $files = array_diff(scandir($dir), ['..', '.']);
+        foreach ($files as $file) {
+            $classname = str_replace('.php', '', $file);
+            $mapLowerToCamelCase[strtolower($classname)] = $classname ;
+        }
+
+        return $mapLowerToCamelCase;
+    }
+
+    public function getClassDefintionByClassId(string $classId, string $tableName): ClassDefinition|null
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/TaskHelper.php]:
```suggestion
    public function getClassDefintionById(string $classId, string $tableName): ClassDefinition|null
```

File: lib/Maintenance/Tasks/TaskHelper.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
+<?php
+declare(strict_types=1);
+
+namespace Pimcore\Maintenance\Tasks;
+
+use Pimcore\Db;
+use Pimcore\Model\DataObject\ClassDefinition;
+use Psr\Log\LoggerInterface;
+
+final class TaskHelper
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/TaskHelper.php]: implement interface here for better testability.

File: lib/Maintenance/Tasks/TaskHelper.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    public function getClassDefintionByClassId(string $classId, string $tableName): ClassDefinition|null
+    {
+        $classDefinition = ClassDefinition::getByIdIgnoreCase($classId);
+        if (!$classDefinition) {
+            $this->logger->error("Classdefinition '" . $classId . "' not found. Please check table " . $tableName);
+        }
+        return $classDefinition;
+    }
+
+    public function cleaningTable(
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/TaskHelper.php]:
```suggestion
    public function cleanupTable(
```
seems more accurate like this..

File: lib/Maintenance/Tasks/DataObject/CleanupBrickTablesTaskHelper.php [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+{
+    private const PIMCORE_OBJECTBRICK_CLASS_DIRECTORY = PIMCORE_CLASS_DIRECTORY . '/DataObject/Objectbrick/Data';
+
+    public function __construct(private LoggerInterface $logger, private DataObjectTaskHelperInterface $helper)
+    {
+    }
+
+    public function cleanupCollectionTable(): void
+    {
+        if (!is_dir(self::PIMCORE_OBJECTBRICK_CLASS_DIRECTORY)) {
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/DataObject/CleanupBrickTablesTaskHelper.php]: I think we should check this in getCollectionNames() as this is required for scandir

File: lib/Maintenance/Tasks/DataObject/DataObjectTaskHelperInterface.php [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+namespace Pimcore\Maintenance\Tasks\DataObject;
+
+/**
+ * @internal
+ */
+interface DataObjectTaskHelperInterface
+{
+    function getCollectionNames(string $dir): array;
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/DataObject/DataObjectTaskHelperInterface.php]: no visibility?

File: lib/Maintenance/Tasks/DataObject/DataObjectTaskHelperInterface.php [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+    function cleanupTable(
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/DataObject/DataObjectTaskHelperInterface.php]: no visibility?

File: lib/Maintenance/Tasks/DataObject/CleanupBrickTablesTaskHelper.php [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+            return;
+        }
+
+        $collectionNames =
+            $this->helper->getCollectionNames(self::PIMCORE_OBJECTBRICK_CLASS_DIRECTORY);
+
+        $db = Db::get();
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/DataObject/CleanupBrickTablesTaskHelper.php]: use DI

File: lib/Maintenance/Tasks/DataObject/CleanupFieldcollectionTablesTaskHelper.php [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    public function cleanupCollectionTable(): void
+    {
+        if (!is_dir(self::PIMCORE_FIELDCOLLECTION_CLASS_DIRECTORY)) {
+            return;
+        }
+
+        $collectionNames =
+            $this->helper->getCollectionNames(self::PIMCORE_FIELDCOLLECTION_CLASS_DIRECTORY);
+
+        $db = Db::get();
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/DataObject/CleanupFieldcollectionTablesTaskHelper.php]: use DI

File: lib/Maintenance/Tasks/DataObject/ConcreteTaskHelperInterface.php [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+namespace Pimcore\Maintenance\Tasks\DataObject;
+
+/**
+ * @internal
+ */
+interface ConcreteTaskHelperInterface
+{
+    function cleanupCollectionTable(): void;
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/DataObject/ConcreteTaskHelperInterface.php]: no visibility?

File: lib/Maintenance/Tasks/DataObject/DataObjectTaskHelper.php [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        bool $isLocalized = true
+    ): void
+    {
+        $classDefinition = ClassDefinition::getByIdIgnoreCase($classId);
+        if (!$classDefinition) {
+            $this->logger->error("Classdefinition '" . $classId . "' not found. Please check table " . $tableName);
+            return;
+        }
+
+        $db = Db::get();
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/DataObject/DataObjectTaskHelper.php]: inject db connection?

File: lib/Maintenance/Tasks/TaskHelper.php [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+use Pimcore\Db;
+use Pimcore\Model\DataObject\ClassDefinition;
+use Psr\Log\LoggerInterface;
+
+final class TaskHelper
+{
+    private static ?TaskHelper $instance = null;
+
+    private LoggerInterface $logger;
+    public static function create(LoggerInterface $logger): self
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/TaskHelper.php]: Is this still used?

File: lib/Maintenance/Tasks/TaskHelper.php [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    public function cleaningTable(
+        string $tableName,
+        ClassDefinition $classDefinition,
+        string $classId,
+        bool $isLocalized = true
+    ): void
+    {
+        $db = Db::get();
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/TaskHelper.php]: inject db connection

File: lib/Maintenance/Tasks/DataObject/CleanupFieldcollectionTablesTaskHelper.php [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+            ],
+        ];
+        foreach ($tasks as $task) {
+            $prefix = $task['prefix'];
+            $pattern = $task['pattern'];
+            $tableNames = $this->db->fetchAllAssociative("SHOW TABLES LIKE '" . $pattern . "'");
+
+            foreach ($tableNames as $tableName) {
+                $tableName = current($tableName);
+                $this->logger->info($tableName . "\n");
```
Comment (dvesh3) [inline lib/Maintenance/Tasks/DataObject/CleanupFieldcollectionTablesTaskHelper.php]:
```suggestion
```
This creates vague entries in the log.

```
[2023-08-01T10:00:19.872032+00:00] app.INFO: object_collection_FilterCategoryMultiselect_EF_FD  [] []
[2023-08-01T10:00:19.916398+00:00] app.INFO: object_collection_FilterCategoryMultiselect_localized_EF_FD  [] []
[2023-08-01T10:00:19.927064+00:00] app.INFO: object_collection_FilterCategory_EF_FD  [] []
[2023-08-01T10:00:19.937106+00:00] app.INFO: object_collection_FilterCategory_localized_EF_FD  [] []
[2023-08-01T10:00:19.947175+00:00] app.INFO: object_collection_FilterInputfield_EF_FD  [] []
[2023-08-01T10:00:19.957478+00:00] app.INFO: object_collection_FilterInputfield_localized_EF_FD  [] []
[2023-08-01T10:00:19.966897+00:00] app.INFO: object_collection_FilterMultiRelation_EF_FD  [] []
[2023-08-01T10:00:19.977077+00:00] app.INFO: object_collection_FilterMultiRelation_localized_EF_FD  [] []
[2023-08-01T10:00:19.986941+00:00] app.INFO: object_collection_FilterMultiSelectFromMultiSelect_EF_FD  [] []
[2023-08-01T10:00:19.996951+00:00] app.INFO: object_collection_FilterMultiSelect_EF_FD  [] []
....-
```

**Review by dvesh3 — APPROVED** [review]

**Pull request merged**